File: Tutorials/Java/7 PublisherConfirms/PublisherConfirms/src/main/java/PublisherConfirms.java
import com.rabbitmq.client.Channel;
import com.rabbitmq.client.ConfirmCallback;
import com.rabbitmq.client.Connection;
import com.rabbitmq.client.ConnectionFactory;
import java.time.Duration;
import java.util.UUID;
import java.util.concurrent.ConcurrentNavigableMap;
import java.util.concurrent.ConcurrentSkipListMap;
import java.util.function.BooleanSupplier;

public class PublisherConfirms {
    private static String host = "localhost";
    private static String user = "guest";
    private static String password = "guest";
    private static final String virtualHost = "/";
    private static int messageCount = 50_000;

    static Connection createConnection() throws Exception {
        ConnectionFactory factory = new ConnectionFactory();
        factory.setHost(host);
        factory.setUsername(user);
        factory.setPassword(password);
        factory.setVirtualHost(virtualHost);

        return factory.newConnection();
    }

    public static void main(String[] args) throws Exception {
        if (args.length > 0) host = args[0];
        if (args.length > 1) user = args[1];
        if (args.length > 2) password = args[2];
        if (args.length > 3) messageCount = Integer.parseInt(args[3]);

        publishMessagesIndividually();
        publishMessagesInBatch();
        handlePublishConfirmsAsynchronously();
    }

    static void publishMessagesIndividually() throws Exception {
        try (Connection connection = createConnection()) {
            Channel channel = connection.createChannel();

            String queue = UUID.randomUUID().toString();
            channel.queueDeclare(queue, false, false, true, null);
            channel.confirmSelect();

            long start = System.nanoTime();
            for (int i = 0; i < messageCount; i++) {
                String body = String.valueOf(i);
                channel.basicPublish("", queue, null, body.getBytes());
                channel.waitForConfirmsOrDie(5_000);
            }
            long end = System.nanoTime();
            System.out.format("Published %,d messages individually in %,d ms%n", messageCount, Duration.ofNanos(end - start).toMillis());
        }
    }

    static void publishMessagesInBatch() throws Exception {
        try (Connection connection = createConnection()) {
            Channel channel = connection.createChannel();

            String queue = UUID.randomUUID().toString();
            channel.queueDeclare(queue, false, false, true, null);
            channel.confirmSelect();

            int batchSize = 100;
            int outstandingMessageCount = 0;

            long start = System.nanoTime();
            for (int i = 0; i < messageCount; i++) {
                String body = String.valueOf(i);
                channel.basicPublish("", queue, null, body.getBytes());
                outstandingMessageCount++;

                if (outstandingMessageCount == batchSize) {
                    channel.waitForConfirmsOrDie(5_000);
                    outstandingMessageCount = 0;
                }
            }

            if (outstandingMessageCount > 0) {
                channel.waitForConfirmsOrDie(5_000);
            }
            long end = System.nanoTime();
            System.out.format("Published %,d messages in batch in %,d ms%n", messageCount, Duration.ofNanos(end - start).toMillis());
        }
    }

    static void handlePublishConfirmsAsynchronously() throws Exception {
        try (Connection connection = createConnection()) {
            Channel channel = connection.createChannel();

            String queue = UUID.randomUUID().toString();
            channel.queueDeclare(queue, false, false, true, null);
            channel.confirmSelect();

            ConcurrentNavigableMap<Long, String> outstandingConfirms = new ConcurrentSkipListMap<>();

            ConfirmCallback cleanOutstandingConfirms = (sequenceNumber, multiple) -> {
                if (multiple) {
                    ConcurrentNavigableMap<Long, String> confirmed = outstandingConfirms.headMap(
                            sequenceNumber, true
                    );
                    confirmed.clear();
                } else {
                    outstandingConfirms.remove(sequenceNumber);
                }
            };

            channel.addConfirmListener(cleanOutstandingConfirms, (sequenceNumber, multiple) -> {
                String body = outstandingConfirms.get(sequenceNumber);
                System.err.format(
                        "Message with body %s has been nack-ed. Sequence number: %d, multiple: %b%n",
                        body, sequenceNumber, multiple
                );
                cleanOutstandingConfirms.handle(sequenceNumber, multiple);
            });

            long start = System.nanoTime();
            for (int i = 0; i < messageCount; i++) {
                String body = String.valueOf(i);
                outstandingConfirms.put(channel.getNextPublishSeqNo(), body);
                channel.basicPublish("", queue, null, body.getBytes());
            }

            if (!waitUntil(Duration.ofSeconds(60), outstandingConfirms::isEmpty)) {
                throw new IllegalStateException("All messages could not be confirmed in 60 seconds");
            }

            long end = System.nanoTime();
            System.out.format("Published %,d messages and handled confirms asynchronously in %,d ms%n", messageCount, Duration.ofNanos(end - start).toMillis());
        }
    }

    static boolean waitUntil(Duration timeout, BooleanSupplier condition) throws InterruptedException {
        int waited = 0;
        while (!condition.getAsBoolean() && waited < timeout.toMillis()) {
            Thread.sleep(100L);
            waited += 100;
        }
        return condition.getAsBoolean();
    }
}

File: Tutorials/Scala/7 PublisherConfirms/src/main/scala/ru/example/rmq/PublisherConfirms.scala
package ru.example.rmq

import com.rabbitmq.client._
import java.time.Duration
import java.util.UUID
import java.util.concurrent.ConcurrentSkipListMap
import java.util.function.BooleanSupplier
import scala.util.Using

object PublisherConfirms {
  private val arguments = new java.util.HashMap[String, Object]()

  def main(args: Array[String]): Unit = {
    val host         = if (args.length > 0) args(0) else "localhost"
    val user         = if (args.length > 1) args(1) else "username"
    val password     = if (args.length > 2) args(2) else "username"
    val messageCount = if (args.length > 3) args(3).toInt else 50_000
    val virtualHost  = "/"

    arguments.put("x-message-ttl", 3600000.asInstanceOf[Object])

    val factory = new ConnectionFactory
    factory.setHost(host)
    factory.setUsername(user)
    factory.setPassword(password)
    factory.setVirtualHost(virtualHost)

    Using.Manager { use =>
      try {
        val connection = use(factory.newConnection)
        val channel    = use(connection.createChannel)
        channel.confirmSelect

        publishMessagesIndividually(channel, messageCount)
        publishMessagesInBatch(channel, messageCount)
        handlePublishConfirmsAsynchronously(channel, messageCount)
      } catch {
        case e: Exception =>
          println(e.getLocalizedMessage)
          sys.exit(-1)
      }
    }
    sys.exit(0)
  }

  private def publishMessagesIndividually(channel: Channel, messageCount: Int): Unit = {
    val queue = UUID.randomUUID.toString
    channel.queueDeclare(queue, false, false, false, arguments)

    val start = System.nanoTime

    (0 until messageCount).foreach { i =>
      channel.basicPublish("", queue, null, i.toString.getBytes)
      channel.waitForConfirmsOrDie(5_000)
    }

    val end = System.nanoTime
    println(s"Published $messageCount messages individually in ${Duration.ofNanos(end - start).toMillis} ms")
  }

  private def publishMessagesInBatch(channel: Channel, messageCount: Int): Unit = {
    val queue = UUID.randomUUID.toString
    channel.queueDeclare(queue, false, false, false, arguments)

    val batchSize = 100
    val quotient  = math.ceil(messageCount / batchSize).toInt
    val remainder = messageCount % batchSize
    val batchList = (0 until batchSize).toList

    val start = System.nanoTime

    (0 until messageCount)
      .zip((0 until quotient).toList.flatMap(_ => batchList) ::: batchList.slice(0, remainder))
      .foreach { ij =>
        channel.basicPublish("", queue, null, ij._1.toString.getBytes)
        if (ij._2 == batchSize) channel.waitForConfirmsOrDie(5_000)
      }
    if (remainder > 0) channel.waitForConfirmsOrDie(5_000)

    val end = System.nanoTime
    println(s"Published $messageCount messages in batch in ${Duration.ofNanos(end - start).toMillis} ms")
  }

  private def handlePublishConfirmsAsynchronously(channel: Channel, messageCount: Int): Unit = {
    val queue = UUID.randomUUID.toString
    channel.queueDeclare(queue, false, false, false, arguments)
    val outstandingConfirms = new ConcurrentSkipListMap[Long, String]

    val cleanOutstandingConfirms: ConfirmCallback = (sequenceNumber: Long, multiple: Boolean) => {
      if (multiple) {
        val confirmed = outstandingConfirms.headMap(sequenceNumber, true)
        confirmed.clear()
      } else outstandingConfirms.remove(sequenceNumber)
    }

    channel.addConfirmListener(
      cleanOutstandingConfirms,
      (sequenceNumber: Long, multiple: Boolean) => {
        val body = outstandingConfirms.get(sequenceNumber)
        System.err.println(
          s"Message with body $body has been nack-ed. Sequence number: $sequenceNumber, multiple: $multiple"
        )
        cleanOutstandingConfirms.handle(sequenceNumber, multiple)
      }
    )

    val start = System.nanoTime

    (0 until messageCount).foreach { i =>
      val body = String.valueOf(i)
      outstandingConfirms.put(channel.getNextPublishSeqNo, body)
      channel.basicPublish("", queue, null, body.getBytes)
    }

    if (!waitUntil(Duration.ofSeconds(60), () => outstandingConfirms.isEmpty))
      throw new IllegalStateException("All messages could not be confirmed in 60 seconds")

    val end = System.nanoTime
    println(s"Published $messageCount messages and handled confirms asynchronously in ${Duration.ofNanos(end - start).toMillis} ms")
  }

  @throws[InterruptedException]
  private def waitUntil(timeout: Duration, condition: BooleanSupplier): Boolean = {
    var waited = 0
    while ({ !condition.getAsBoolean && waited < timeout.toMillis }) {
      Thread.sleep(100L)
      waited = +100
    }
    condition.getAsBoolean
  }
}
